controllers/nonadmin_controller.go

```diff
@@ -0,0 +1,193 @@
+)
+
+const (
+	nonAdminPrefix            = "non-admin-"
+	controllerManager         = "controller-manager"
+	containerName             = nonAdminPrefix + "manager"
+	nonAdminControllerManager = common.OADPOperatorPrefix + nonAdminPrefix + controllerManager
+	controlPlaneKey           = "control-plane"
+)
+
+var (
+	controlPlaneLabel = map[string]string{
+		controlPlaneKey: nonAdminPrefix + controllerManager,
+	}
+	deploymentLabels = map[string]string{
+		"app.kubernetes.io/component":  "manager",
+		"app.kubernetes.io/created-by": common.OADPOperator,
+		"app.kubernetes.io/instance":   nonAdminPrefix + controllerManager,
+		"app.kubernetes.io/managed-by": "kustomize",
+		"app.kubernetes.io/name":       "deployment",
+		"app.kubernetes.io/part-of":    common.OADPOperator,
+	}
+)
+
+func (r *DPAReconciler) ReconcileNonAdminController(log logr.Logger) (bool, error) {
+	// TODO https://github.com/openshift/oadp-operator/pull/1316
+	dpa := oadpv1alpha1.DataProtectionApplication{}
+	if err := r.Get(r.Context, r.NamespacedName, &dpa); err != nil {
+		return false, err
+	}
+
+	nonAdminDeployment := &appsv1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      nonAdminControllerManager,
+			Namespace: r.NamespacedName.Namespace,
+		},
+	}
+
+	// Delete (possible) previously deployment
+	if !r.checkNonAdminEnabled(&dpa) {
+		if err := r.Get(
+			// TODO use r.Context?
+			context.Background(),
+			types.NamespacedName{
+				Name:      nonAdminDeployment.Name,
+				Namespace: nonAdminDeployment.Namespace,
+			},
+			nonAdminDeployment,
+		); err != nil {
+			if k8serror.IsNotFound(err) {
+				return true, nil
+			}
+			return false, err
+		}
+
+		deleteOptionPropagationForeground := metav1.DeletePropagationForeground
+		if err := r.Delete(
+			// TODO use r.Context?
+			context.Background(),
+			nonAdminDeployment,
+			&client.DeleteOptions{PropagationPolicy: &deleteOptionPropagationForeground},
+		); err != nil {
+			r.EventRecorder.Event(
+				nonAdminDeployment,
+				corev1.EventTypeWarning,
+				"NonAdminDeploymentDeleteFailed",
+				fmt.Sprintf("Could not delete non admin controller deployment %s/%s: %s", nonAdminDeployment.Namespace, nonAdminDeployment.Name, err),
+			)
+			return false, err
+		}
+		r.EventRecorder.Event(
+			nonAdminDeployment,
+			corev1.EventTypeNormal,
+			"NonAdminDeploymentDeleteSucceed",
+			fmt.Sprintf("Non admin controller deployment %s/%s deleted", nonAdminDeployment.Namespace, nonAdminDeployment.Name),
+		)
+		return true, nil
+	}
+
+	operation, err := controllerutil.CreateOrUpdate(
+		r.Context,
+		r.Client,
+		nonAdminDeployment,
+		func() error {
+			nonAdminImage := r.getNonAdminImage(&dpa)
+			if len(nonAdminImage) == 0 {
+				return fmt.Errorf("no Non Admin Controller image found in RELATED_IMAGE_NON_ADMIN_CONTROLLER environment variable or unsupportedOverrides")
+			}
+			ensureRequiredLabels(nonAdminDeployment)
+			ensureRequiredSpecs(nonAdminDeployment, nonAdminImage)
```

Comment:
the way this is implemented, it does not allow user/developer to change parameters of the deployment, like replicas. Is this ok?

Reply:
yes this seems ok, but we may need some more customization of deployment/container in the very near future. so What do you think about this:
- First, lets remove the `nonAdminImage`parameter from the `ensureRequiredSpecs` function and fetch that image value inside the `ensureRequiredSpecs` function and use it
- Rename the `ensureRequiredSpecs` function to `buildNonAdminDeployment`

Reply:
done